test/unit/signal.spec.js

```diff
@@ -1,107 +1,152 @@
-import Signal from '../../lib/api/signal'
+import { MemoizedSignal, Signal } from '../../lib/api/signal'
 import {noop} from '../helpers'
 
-describe(`Signal`, () => {
-  describe(`constructor`, () => {
-    it(`creates a Signal instance`, () => {
-      expect(new Signal()).to.be.instanceof(Signal)
-    })
-  })
 
-  describe(`instance`, () => {
-    let signal
+function runCommonSpecs (SignalConstructor, initialValue = 42) {
+  describe('instance', () => {
+    let signalInstance
+    let spies
     beforeEach(() => {
-      signal = new Signal()
+      signalInstance = new SignalConstructor(initialValue)
+      spies = {
+        one: sinon.spy(),
+        two: sinon.spy(),
+        three: sinon.spy(),
+        reset () {
+          this.one.reset()
+          this.two.reset()
+          this.three.reset()
+        }
+      }
     })
 
-    describe(`attach(listener)`, () => {
-      it(`returns a function`, () => {
-        expect(signal.attach(noop)).to.be.a('function')
+    describe('attach(listener)', () => {
+      it('returns a function', () => {
+        expect(signalInstance.attach(noop)).to.be.a('function')
       })
-      it(`throws an error if listener is not a function`, () => {
+
+      it('throws an error if listener is not a function', () => {
         ['foo', undefined, 42].forEach((value) => {
-          expect(() => { signal.attach(value) }).to.throw()
+          expect(() => { signalInstance.attach(value) }).to.throw()
         })
       })
-    })
 
-    describe(`dispatch()`, () => {
-      it(`returns nothing and does not fail`, () => {
-        expect(signal.dispatch()).to.equal(undefined)
+      it('detaches listener when returned function is invoked', () => {
+        const detach = signalInstance.attach(spies.one)
+
+        spies.reset() // deal with memoization behaviour
+
+        signalInstance.dispatch('holdthedoor')
+        sinon.assert.calledOnce(spies.one)
+        spies.reset()
+        detach()
+        signalInstance.dispatch()
+        sinon.assert.notCalled(spies.one)
       })
+    })
 
-      describe(`with previously attached listeners`, () => {
-        let spies
-        beforeEach(() => {
-          spies = {
-            one: sinon.spy(),
-            two: sinon.spy(),
-            three: sinon.spy()
-          }
-        })
+    describe('dispatch()', () => {
+      it('returns nothing and does not fail', () => {
+        expect(signalInstance.dispatch()).to.equal(undefined)
+      })
 
+      describe('with previously attached listeners', () => {
         function expectCallCount (obj) {
```

Comment:
Can we make this a method on spies. Otherwise it is not very clear that it depends on the state of spies.